bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java

```diff
@@ -0,0 +1,949 @@
+        }
+
+        InetAddress addr = InetAddress.getByName(handler.getAddress());
+        Integer port = handler.getPort();
+
+        if (port != null) {
+            Socket socket = new Socket(addr, port);
+            qSocket = socket;
+            qOut = new PrintWriter(socket.getOutputStream(), true);
+            qIn = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+            logger.debug("Connected via local port {} from thread {}", socket.getLocalPort(),
+                    Thread.currentThread().getId());
+        } else {
+            return;
+        }
+
+        setSN();
+        getSN();
+
+        // Connect to Qbus server
+        connect();
+
+        // If Qbus Client is connected then initialize and start listening for incoming commands, else put Bridge
+        // offline
+        if (ctdConnected) {
+            initialize();
+            threadExecutor.execute(() -> {
+                try {
+                    qEvents();
+                } catch (IOException e) {
+                    logger.debug("Could not start thread {} ", e.getMessage());
+                }
+            });
+        } else {
+            handler.bridgeOffline("No communication with Qbus client");
+        }
+    }
+
+    /**
+     * Cleanup socket when the communication with Qbus Server is closed.
+     *
+     * @throws IOException
+     *
+     */
+    public synchronized void stopCommunication() throws IOException {
+        listenerStopped = true;
+
+        Socket socket = qSocket;
+
+        if (socket != null) {
+            try {
+                socket.close();
+            } catch (IOException ignore) {
+                // ignore IO Error when trying to close the socket if the intention is to close it anyway
+            }
+        }
+
+        qSocket = null;
+
+        PrintWriter qbusOut = this.qOut;
+        if (qbusOut != null) {
+            qbusOut.close();
+        }
+
+        BufferedReader qbusIn = this.qIn;
+        try {
+            if (qbusIn != null) {
+                qbusIn.close();
+            }
+        } catch (IOException e) {
+            logger.debug("Error on closing reader {} ", e.getMessage());
+        }
+
+        ctdConnected = false;
+        logger.debug("Communication stopped from thread {}", Thread.currentThread().getId());
+    }
+
+    /**
+     * Close and restart communication with Qbus Server.
+     *
+     * @throws InterruptedException
+     * @throws IOException
+     */
+
+    public synchronized void restartCommunication() throws InterruptedException, IOException {
+        stopCommunication();
+
+        logger.debug("Qbus: restart communication from thread {}", Thread.currentThread().getId());
+
+        try {
+            startCommunication();
+        } catch (InterruptedException | IOException e) {
+            String message = e.toString();
+            logger.error("Could not start the communication with the server. {}", message);
```

Comment:
Bindings should only log to error if something severe happened, like the detection of a bug in your code. This could be debug. See this link for a description of the log levels: https://www.openhab.org/docs/developer/guidelines.html#f-logging